### ISSUE TITLE
pods cannot be scheduled, bug nodes has enough resources

### ISSUE THREAD

**Issue by attlee-wang** [kubernetes/kubernetes#86626]
**What happened**:
I have a large k8s cluster with 500~600 nodes(The node specifications and label are exactly the same) and autoscaler using running a lot of jobs(jobs also the same). I found a problem that some pods cannot be scheduled and always in the pending status, bug nodes has enough resources. that node scheduled pod at once when restart kube-scheduler.  I suspect that's because kueb-scheduler cache and kube-api cache are inconsistent, so I extended the scheduler debug interface to troubleshoot this problem. But I didn't find any useful information.
as：
pod pending list：
![image](https://user-images.githubusercontent.com/57479557/71455050-515cbd80-27ce-11ea-90d5-702be5b5d082.png)
![image](https://user-images.githubusercontent.com/57479557/71455367-9c2b0500-27cf-11ea-9c4b-fa3c2ea2ec18.png)

describe pod:
![image](https://user-images.githubusercontent.com/57479557/71455098-81a45c00-27ce-11ea-8aed-7e2558898997.png)

describe node:
![image](https://user-images.githubusercontent.com/57479557/71455119-9a147680-27ce-11ea-9e69-79e4a968f9ce.png)

**What you expected to happen**:
Because there are some pods that has been pending for a long time, so autoscaler not working. I expected that nodes scheduled with nodes has enough resources.

**How to reproduce it (as minimally and precisely as possible)**:
The difficulty of this problem is that it is difficult to reproduce it. I tried many times, but it didn't reproduce and like a probability. But it's really appeared in our online cluster and more nodes,more likely.

**Anything else we need to know?**:
I wrote a shell script using check some node not scheduled pod in 30min：
```
#!/bin/bash -e

##
workerDir=~/monitor
schedulerIP=127.0.0.1
cluster=clusterId
namespeace=default
podKey=worker
podStatus=Completed
sleepTime=10m

echo schedulerIP:$schedulerIP cluster:$cluster namespeace:$namespeace podKey:$podKey podStatus:$podStatus sleepTime:$sleepTime > result
## init
cd $workerDir
>  0min
> 10min
> 20min
> 30min
count=0

##
while :
do
        ##
        #echo =========nodeAll===============
        kubectl get node -o wide | egrep -v "NAME" | awk '{print $1}' > nodeAll
        #cat nodeAll

        #echo =========podOnNode=============
        kubectl get pod -n $namespeace -o wide| grep $podKey | egrep -v $podStatus | egrep -v 'NAME' | awk '{print $7}' > podOnNode
        #cat nodeAll

        #echo ========nodeAll-podOnNode======
        sort -m <(sort nodeAll | uniq) <(sort podOnNode | uniq) <(sort podOnNode | uniq) | uniq -u > nodeAll-podOnNode        
        #cat nodeAll-podOnNode

        #echo ========exceptNode=============
        #cat exceptNode

        #echo ===========one-result==========
        sort -m <(sort nodeAll-podOnNode | uniq) <(sort exceptNode | uniq) <(sort exceptNode| uniq) | uniq -u > 0min
        #cat 0min

        ##
        cat 20min > 30min
        cat 10min > 20min
        cat 0min  > 10min

        ##
        #echo ============result=============
        count=$(($count+1))
        echo ========== check-count:$count ========= >> result
        nodelist=`sort -m <(sort 0min | uniq) <(sort 30min | uniq) | uniq -d`
        echo ">>>> exit node not scheduled: " $nodelist >> result

       ## The following interface that I extended the scheduler to troubleshoot this problem. But I didn't find any useful information.
        for node in $nodelist; do  
           echo ">>>> scheduler cached node: $node info<<<<" >> result
           echo `curl http://$schedulerIP:10251/debug/node/cached?hostname=$node >> result`
           echo -e '\n' >> result

           echo ">>>> kube-api node: $node info<<<<" >> result
           echo `curl http://$schedulerIP:10251/debug/api/node/infos?hostname=$node >> result`
           echo -e '\n' >> result

           echo ">>>> scheduler cached all nodes resource info<<<<" >> result
           echo `curl http://$schedulerIP:10251/debug/nodes/resource >> result`
           echo -e '\n' >> result

           echo ">>>>kube-api node: $node resource info<<<<" >> result
           echo `curl http://$schedulerIP:10251/debug/api/node/resource?hostname=$node >> result`
           echo -e '\n' >> result
        done
        echo -e '\n\n' >> result

        #cat result
        sleep $sleepTime
done
``` 

**Environment**:  go version go1.12.13 linux/amd64
- Kubernetes version (use `kubectl version`):  v1.12.4
- Cloud provider or hardware configuration: TKE
- OS (e.g: `cat /etc/os-release`): ubuntu18.04
- Kernel (e.g. `uname -a`):  
Linux VM_227_154_centos 3.10.107-1-tlinux2_kvm_guest-0047 #1 SMP Mon Oct 22 15:37:31 CST 2018 x86_64 x86_64 x86_64 GNU/Linux

**Comment by attlee-wang** [comment]
@liggitt Look at the problem

**Comment by attlee-wang** [comment]
@liggitt See at the problem

**Comment by liggitt** [comment]
Is this reproducible against a 1.15+ version of kubernetes?

**Comment by liggitt** [comment]
@kubernetes/sig-scheduling-bugs

**Comment by attlee-wang** [comment]
@liggitt I'm not sure, except for v1.12.4, I didn't find this problem in other versions. Even in version v1.12.4, I tried many times and didn't  reproduce it. It's difficult to reproduce it. If you need other information, you can @ me.

**Comment by attlee-wang** [comment]
@wojtek-t @fgrzadkowski @gmarek @mwielgus See at the problem，it is similar to [#34772](https://github.com/kubernetes/kubernetes/issues/34772) and  [#37232](https://github.com/kubernetes/kubernetes/issues/37232). It seems that pod has been assumed on node, so other pod not scheduled. it’s ok when restart kube-scheduler. This is just my guess. and It needs to be confirmed.

**Comment by attlee-wang** [comment]
@honkiko

**Comment by wojtek-t** [comment]
https://github.com/kubernetes/kubernetes/pull/73845 might have been the culprit - and that was only cherrypicked in 1.12.6

**Comment by snowplayfire** [comment]
@liggitt @wojtek-t @attlee-wang I meet the same problem in v1.13.4.

There are more than four scenarios as follows:

1. pod is pending when node has enough resources

2. pod is scheduled into the node which hasn't enough resources

3. pod is scheduled into the node which is cordoned

4. pod is scheduled into the node which has new taint

The problem can be solved temporarily by two ways:

1. change kube-scheduler's leader

or

2. delete the node, re-add it to the cluster 

That is, kube-scheduler's cache is inconsistent with apiserver under some conditions. But I don't know how to reproduce it.

**Comment by attlee-wang** [comment]
@wojtek-t how to reproduce #73845 issues before accepted it？I want to make sure #73845 is my problem.

**Comment by praseodym** [comment]
> @liggitt @wojtek-t @attlee-wang I meet the same problem in v1.13.2.

#73845 was cherrypicked in v1.13.4, so v1.13.2 is still susceptible to that bug.

**Comment by snowplayfire** [comment]
@wojtek-t @praseodym 
I read the PR #73845, but I' m confused and have a question.
As #73845,  these "repeated" events may appear when a watcher starts. That is, scheduler can not schedule pods even though they fit on a node only when scheduler restarts.

However,  I meet the problem not when scheduler restarts.  

My scenario:
Scheduler works well for a period of time, then this problem occurs.  But only the leader scheduler has the problem,  the problem can disappear after I change  the scheduler's  leader(for example: stop the leader)

Am I wrong？  Hope for your help.

**Comment by wojtek-t** [comment]
No - that has nothing to do with scheduler restarts.
This might have caused problems on every watch restart, which happens every 5-10 minutes...

**Comment by attlee-wang** [comment]
@wojtek-t  But about this issue, I'm quite sure that node scheduled pod at once when restart kube-scheduler. How to explain?

**Comment by snowplayfire** [comment]
@wojtek-t  @praseodym  Sorry to find that our scheduler version includes #73845, but still has the problem.

**Comment by snowplayfire** [comment]
@wojtek-t  @praseodym  Sorry to find that our scheduler version includes #73845, but still has the problem.

**Comment by snowplayfire** [comment]
https://github.com/kubernetes/kubernetes/issues/86986  Kubelet may have the similar problem.

**Comment by alculquicondor** [comment]
@wojtek-t can something be done about such old versions in kube-apiserver if #73845 is the culprit?

@snowplayfire @attlee-wang could you try to repro in 1.16+?

**Comment by wojtek-t** [comment]
> @wojtek-t can something be done about such old versions in kube-apiserver if #73845 is the culprit?

restarting scheduler would help

**Comment by alculquicondor** [comment]
I meant to ask if we can patch and fix the underlying issue. I'm inclined to close this issue if we can't repro in 1.16+

**Comment by wojtek-t** [comment]
Hmm - are you asking about the correct PR? It was merged in 1.14 and cherrypicked to 1.11-1.13...

**Comment by alculquicondor** [comment]
> #73845 might have been the culprit - and that was only cherrypicked in 1.12.6

Then you meant this was the fix, rather than the culprit? That would make more sense given than OP is running v1.12.4

Although @snowplayfire says the bug is still there for them. Could you confirm the version you are running?
